prepreduce/prepreduce.py

```diff
@@ -0,0 +1,395 @@
+              {make_result.stdout=}
+              {cc=}
+              {local_path_for_dot_i_file=}
+              {make_result.returncode}
+              """
+        )
+
+    success(
+        f"Made the .i target file successfully. It's located here: {local_path_for_dot_i_file}"
+    )
+
+    return local_path_for_dot_i_file
+
+
+def get_compiler_invocation(make_stdout: str) -> list[str]:
+    """
+    Parse out the CC invocation from make's raw output with V=1 on
+    """
+    pattern = r".*-c -o.*"  # HACK: how robust is this pattern?
+    matches = re.findall(pattern, make_stdout)
+    ic(matches)
+    assert (
+        len(matches) == 1
+    ), "Too many (or too few) compiler invocation matches! (not your fault, it's prepreduce's fault)"
+
+    main_invocation = matches[0].split()
+    return main_invocation
+
+
+def get_clean_compiler_invocation(cc_invocation: list[str]) -> list[str]:
+    """
+    Remove all preprocessor flags as well as any relative pathing in invocation
+    """
+    rules = [
+        lambda flag: flag[:2] != "-I",
+        lambda flag: flag[:3] != "-Wp",
+        lambda flag: flag[:2] != "-D",
+        lambda flag: flag != "-include",
+        lambda flag: r"./" not in flag,  # no relative pathing allowed!
+        lambda flag: not flag.startswith("-Werror"),
+    ]
+
+    cleaned = cc_invocation[::]
+    for rule in rules:
+        cleaned = filter(rule, cleaned)
+
+    cleaned = list(cleaned)
+
+    ic(cleaned)
+
+    return cleaned
+
+
+def main() -> None:
+    cli_args: argparse.Namespace = parse_cli_args()
+
+    ic(cli_args)
+    make_stdout = initial_make(
+        build_command=cli_args.build_command,
+        target=cli_args.target,
+        path_to_linux=cli_args.path_to_linux,
+        force_rm=cli_args.force_rm_existing_target,
+    )
+
+    ic(make_stdout)
+
+    cc_invocation = get_compiler_invocation(make_stdout)
+    path_to_dot_i_file = make_dot_i_file(
+        cc_invocation, path_to_linux=cli_args.path_to_linux, output_dir=cli_args.output
+    )
+
+    clean_cc_invocation = get_clean_compiler_invocation(cc_invocation)
+    clean_cc_invocation[-1] = str(
+        path_to_dot_i_file
+    )  # HACK: is last arg always target? i.e: `lib/xyz.c`
+    ic(clean_cc_invocation)
+
+    write_test_script(
+        cc_invocation=clean_cc_invocation,
+        output_dir=cli_args.output,
+        target=path_to_dot_i_file.with_suffix(".o"),
+        force_rm=cli_args.force_rm_existing_script,
+    )
+
+
+def write_test_script(
+    *, cc_invocation: list[str], output_dir: Path, target: Path, force_rm: bool
+) -> None:
+    """
+    Create a test.sh script at --output directory
+    """
+    script_path = Path(os.getcwd()) / Path("test.sh")
+    info(f"Generating {script_path}")
+    if script_path.exists():
+        double_check_removal_with_user(file=script_path, force_rm=force_rm)
+
+    output_target_idx = cc_invocation.index("-o") + 1
+    cc_invocation[output_target_idx] = str(target)
```

Comment:
As mentioned before, `subprocess.run()` can handle path-like objects, so the cast to `str` should not be necessary.